mathics/builtin/base.py

```diff
@@ -728,10 +728,7 @@ def __init__(self, name, tag, *args):
 
 class PatternArgumentError(PatternError):
     def __init__(self, name, count, expected):
-        super().__init__(None, None)
-        self.name = name
-        self.count = count
-        self.expected = expected
+        super().__init__(name, "argr", count, expected)
```

Comment:
I probably wasn't clear. In `super()` only `name`, and `count` are saved in `self`, Although `expected` is accepted, nothing is done with it (and therefore there is no benefit in passing it as an argument. In fact it can lead to confusion suggesting something might be done with it).

In particular, `expected` is not stored into `self`, right? 

If that's the case, then you still would need a `self.expected = `expected` if you expect this value to be set.

Reply:
I still have to work out better how these exceptions and warnings are handled. The current implementation in main is not entirely correct, and this one neither...

Reply:
```
$ mathics

Mathics 2.1.1.dev0
on CPython 3.6.13 (default, Feb 23 2021, 10:08:17) 
using SymPy 1.8, mpmath 1.2.1, numpy 1.19.5, cython 0.29.22

Copyright (C) 2011-2021 The Mathics Team.
This program comes with ABSOLUTELY NO WARRANTY.
This is free software, and you are welcome to redistribute it
under certain conditions.
See the documentation for the full license.

Quit by evaluating Quit[] or by pressing CONTROL-D.

In[1]:= MatchQ[3, Pattern[3]]
Traceback (most recent call last):
  File "/home/rocky/.pyenv/versions/3.6.13/bin/mathics", line 11, in <module>
    load_entry_point('Mathics3', 'console_scripts', 'mathics')()
  File "/src/external-vcs/github/mathics/Mathics/mathics/main.py", line 369, in main
    result = evaluation.evaluate(query, timeout=settings.TIMEOUT)
  File "/src/external-vcs/github/mathics/Mathics/mathics/core/evaluation.py", line 355, in evaluate
    result = run_with_timeout_and_stack(evaluate, timeout, self)
  File "/src/external-vcs/github/mathics/Mathics/mathics/core/evaluation.py", line 112, in run_with_timeout_and_stack
    return request()
  File "/src/external-vcs/github/mathics/Mathics/mathics/core/evaluation.py", line 326, in evaluate
    self.last_eval = query.evaluate(self)
  File "mathics/core/expression.py", line 1310, in mathics.core.expression.Expression.evaluate
    expr, reevaluate = expr.evaluate_next(evaluation)
  File "mathics/core/expression.py", line 1444, in mathics.core.expression.Expression.evaluate_next
    result = rule.apply(new, evaluation, fully=False)
  File "mathics/core/rules.py", line 67, in mathics.core.rules.BaseRule.apply
    self.pattern.match(yield_match, expression, {}, evaluation, fully=fully)
  File "mathics/core/pattern.py", line 291, in mathics.core.pattern.ExpressionPattern.match
    self.head.match(yield_head, expression.get_head(), vars, evaluation)
  File "mathics/core/pattern.py", line 159, in mathics.core.pattern.AtomPattern.match
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 281, in mathics.core.pattern.ExpressionPattern.match.yield_head
    self.get_pre_choices(
  File "mathics/core/pattern.py", line 429, in mathics.core.pattern.ExpressionPattern.get_pre_choices
    yield_func(vars)
  File "mathics/core/pattern.py", line 258, in mathics.core.pattern.ExpressionPattern.match.yield_choice
    self.match_leaf(
  File "mathics/core/pattern.py", line 655, in mathics.core.pattern.ExpressionPattern.match_leaf
    self.get_wrappings(
  File "mathics/core/pattern.py", line 456, in mathics.core.pattern.ExpressionPattern.get_wrappings
    yield_func(items[0])
  File "mathics/core/pattern.py", line 643, in mathics.core.pattern.ExpressionPattern.match_leaf.yield_wrapping
    leaf.match(
  File "mathics/builtin/patterns.py", line 829, in mathics.builtin.patterns.Pattern_.match
    self.pattern.match(yield_func, expression, new_vars, evaluation)
  File "mathics/builtin/patterns.py", line 1027, in mathics.builtin.patterns.Blank.match
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 623, in mathics.core.pattern.ExpressionPattern.match_leaf.match_yield
    self.match_leaf(
  File "mathics/core/pattern.py", line 655, in mathics.core.pattern.ExpressionPattern.match_leaf
    self.get_wrappings(
  File "mathics/core/pattern.py", line 456, in mathics.core.pattern.ExpressionPattern.get_wrappings
    yield_func(items[0])
  File "mathics/core/pattern.py", line 643, in mathics.core.pattern.ExpressionPattern.match_leaf.yield_wrapping
    leaf.match(
  File "mathics/builtin/patterns.py", line 829, in mathics.builtin.patterns.Pattern_.match
    self.pattern.match(yield_func, expression, new_vars, evaluation)
  File "mathics/builtin/patterns.py", line 1027, in mathics.builtin.patterns.Blank.match
    yield_func(vars, None)
  File "mathics/core/pattern.py", line 640, in mathics.core.pattern.ExpressionPattern.match_leaf.match_yield
    yield_func(new_vars, items_rest)
  File "mathics/core/pattern.py", line 616, in mathics.core.pattern.ExpressionPattern.match_leaf.leaf_yield
    yield_func(
  File "mathics/core/rules.py", line 42, in mathics.core.rules.BaseRule.apply.yield_match
    new_expression = self.do_replace(expression, vars, options, evaluation)
  File "mathics/core/rules.py", line 131, in mathics.core.rules.BuiltinRule.do_replace
    return self.function(evaluation=evaluation, **vars_noctx)
  File "mathics/builtin/patterns.py", line 670, in mathics.builtin.patterns.MatchQ.apply
    if match(expr, form, evaluation):
  File "mathics/builtin/patterns.py", line 647, in mathics.builtin.patterns.match
    return Matcher(form).match(expr, evaluation)
  File "mathics/builtin/patterns.py", line 633, in mathics.builtin.patterns.Matcher.__init__
    self.form = Pattern.create(form)
  File "mathics/core/pattern.py", line 24, in mathics.core.pattern.Pattern_create
    return pattern_object(expr)
  File "/src/external-vcs/github/mathics/Mathics/mathics/builtin/base.py", line 382, in __new__
    instance.init(*args, **kwargs)
  File "mathics/builtin/patterns.py", line 804, in mathics.builtin.patterns.Pattern_.init
    super(Pattern_, self).init(expr)
  File "/src/external-vcs/github/mathics/Mathics/mathics/builtin/base.py", line 740, in init
    self.error_args(len(expr.leaves), *self.arg_counts)
  File "/src/external-vcs/github/mathics/Mathics/mathics/builtin/base.py", line 749, in error_args
    raise PatternArgumentError(self.get_name(), count, *expected)
mathics.builtin.base.PatternArgumentError
````